src/impl_/coroutine.rs

```diff
@@ -0,0 +1,19 @@
+use crate::coroutine::Coroutine;
+use crate::impl_::wrap::OkWrap;
+use crate::{IntoPy, PyErr, PyObject, Python};
+use std::future::Future;
+
+/// Used to wrap the result of async `#[pyfunction]` and `#[pymethods]`.
+pub fn wrap_future<F, R, T>(future: F) -> Coroutine
+where
+    F: Future<Output = R> + Send + 'static,
+    R: OkWrap<T> + Send,
+    T: IntoPy<PyObject> + Send,
+    R::Error: Send,
+    PyErr: From<R::Error>,
+{
+    Coroutine::from_future(async move {
+        // SAFETY: GIL is acquired when future is polled (see `Coroutine::poll`)
+        future.await.wrap(unsafe { Python::assume_gil_acquired() })
```

Comment:
Can this transformation be moved inside `from_future`, to put the safety invariant closer to `Coroutine::poll` ?

Reply:
I wanted to do so, but it requires to expose `OkWrap` trait. If you are ok with that move, I can update `from_future` signature.